### ISSUE TITLE
Add more optimisers!

### ISSUE THREAD

**Issue by samuel-watson** [samuel-watson/glmmrBase#34]
- [ ] NEWUOA: There's a c++ implementation in this library: https://github.com/elsid/newuoa-cpp but also look at other libraries.
- [ ] LBFGS: https://github.com/ZJU-FAST-Lab/LBFGS-Lite/tree/master

**Comment by zaikunzhang** [comment]
Hi @samuel-watson,

If you are interested in NEWUOA and other derivative-free solvers by the late Professor M.J.D. Powell, I suggest you have a look at [PRIMA](http://www.libprima.net): 

https://github.com/libprima/prima

The implementation you mentioned is based on the Fortran 77 version of NEWUOA, which is buggy, unmaintained, and should not be used any more. 

Thanks.

Zaikun

**Comment by samuel-watson** [comment]
Hi @zaikunzhang Thanks so much for the pointer - I'll use the implementation you point to. Ideally I'm looking for derivative-free solvers that are competitive with BOBYQA for minimising convex functions - nothing seems to beat BOBYQA for the problems I have tried but NEWUOA looked promising.

**Comment by zaikunzhang** [comment]
PRIMA provides BOBYQA as well, in addition to NEWUOA, UOBYQA, LINCOA, and COBYLA. 

The PRIMA implementation of BOBYQA performs much better than any version based on the original Fortran 77 implementation in terms of number of function evaluations.

**Comment by samuel-watson** [comment]
That's interesting @zaikunzhang . We are currently using [this implementation](https://github.com/samuel-watson/glmmrBase/blob/master/inst/include/glmmr/optim/bobyqa_general.h), which looks like its based on the old Fortran version - so I'll update that as well.

**Comment by zaikunzhang** [comment]
> That's interesting @zaikunzhang . We are currently using [this implementation](https://github.com/samuel-watson/glmmrBase/blob/master/inst/include/glmmr/optim/bobyqa_general.h), which looks like its based on the old Fortran version - so I'll update that as well.

Yes, I am sure it is, considering the GOTO's. Indeed, PRIMA is the first modernized implementation of Powell's solvers, in the sense that it is totally GOTO-free and makes use of matrix-vector procedures.

**Comment by samuel-watson** [comment]
Hi @zaikunzhang I am just looking at prima's implementation of BOBYQA for use in C++ program. In my case the number of parameters is a run-time parameter, so it looks like I'd need to dynamically allocate an array and then pass that. Ideally, I'd like to pass a pointer to the start of the data, such as the start of a std vector - do you know if this is possible with the prima code, or do I need a dynamic array?

**Comment by samuel-watson** [comment]
Hi @zaikunzhang this looks great. My knowledge of Fortran is basically nonexistent though so even with the C bindings I'm not sure how to incorporate it. The problem is this package is mainly written in C++ to then be used in a package for R. Unfortunately R's package repository CRAN is really strict about things to include in package, so you couldn't incorporate a dll or anything pre-built and it looks like you have to build the fortran program first and then link to the C program. I'll try to figure this out in the future. It'd be great if this could eventually be used with R's C++ interface as there are a huge number of statistical applications this would be great for.

**Comment by zaikunzhang** [comment]
Hi @samuel-watson ,

It would be great if PRIMA can be bonded to R. Although I am ignorant of R, I will be very glad to provide assistance from the Fortran side.

For your information, see

https://github.com/lme4/lme4/issues/744

https://github.com/libprima/prima/issues/104

**Comment by samuel-watson** [comment]
Thanks @zaikunzhang - interesting to see that they are looking at it for lme4. For R there are two ways that are of interest: making the fortran functions available as calls in R, and making the fortran code callable from C++ for R (with Rcpp). I'm interested in the latter as a lot of useful code is written in C++ like this package. It is possible to use both in R, see [here](https://gallery.rcpp.org/articles/Combining-R-with-Cpp-and-Fortran/). However, that example has a simple fortran function. The problem I have is that I am unfamiliar with fortran, and given the size of the prima library, I don't know if it will be as straightforward as doing something like:
```
extern "C" {
    double bobyqa_(int *n, double *x, ...);
}
```
and then calling the `bobyqa_` function in C++.  I don't know if you have any insights on this? The other problem is making it compliant with the (frustrating) CRAN standards.

**Comment by samuel-watson** [comment]
@zaikunzhang Your comments have me thinking whether it would be worth developing a native C++ implementation of these algorithms using modern C++ and incorporating your improvements to the algorithms. Certainly a lot of work, but would be interesting to translate your project as a header only library - would give me an opportunity to learn fortran!

**Comment by zaikunzhang** [comment]
Hi @samuel-watson ,

> Certainly a lot of work, but would be interesting to translate your project as a header only library. I'd be interested in giving this a go if that would be of interest. I can open an issue on the prima repo

Certainly! This is highly interesting. Indeed, just this morning I was thinking about the native C or C++ implementation of PRIMA. However, my knowledge about C/C++ is limited and outdated, so I cannot do that myself. It would be great if you could do it. 

Thank you very much!
